src/interfaces/IEarnerManager.sol

```diff
@@ -0,0 +1,178 @@
+// SPDX-License-Identifier: BUSL-1.1
+
+pragma solidity 0.8.26;
+
+import { IMigratable } from "../../lib/common/src/interfaces/IMigratable.sol";
+
+/**
+ * @title  Earner Status Manager interface for setting and returning earner status for Wrapped M Token accounts.
+ * @author M^0 Labs
+ */
+interface IEarnerManager is IMigratable {
+    /* ============ Events ============ */
+
+    /**
+     * @notice Emitted when the earner for `account` is set to `status`.
+     * @param  account The account under which yield could generate.
+     * @param  status  Whether the account is set as an earner, according to the admin.
+     * @param  admin   The admin who set the details and who will collect the fee.
+     * @param  feeRate The fee rate to be taken from the yield.
+     */
+    event EarnerDetailsSet(address indexed account, bool indexed status, address indexed admin, uint16 feeRate);
+
+    /* ============ Custom Errors ============ */
+
+    /// @notice Emitted when `account` is already in the earners list, so it cannot be added by an admin.
+    error AlreadyInRegistrarEarnersList(address account);
+
+    /// @notice Emitted when the lengths of input arrays do not match.
+    error ArrayLengthMismatch();
+
+    /// @notice Emitted when the length of an input array is 0.
+    error ArrayLengthZero();
+
+    /// @notice Emitted when the earner details have already be set by an existing and active admin.
+    error EarnerDetailsAlreadySet(address account);
+
+    /// @notice Emitted when the earners lists are ignored, thus not requiring admin to define earners.
+    error EarnersListsIgnored();
+
+    /// @notice Emitted when the fee rate provided is to high (higher than 100% in basis points).
+    error FeeRateTooHigh();
+
+    /// @notice Emitted when setting fee rate to a nonzero value while setting status to false.
+    error InvalidDetails();
+
+    /// @notice Emitted when the caller is not an admin.
+    error NotAdmin();
+
+    /// @notice Emitted when the non-governance migrate function is called by a account other than the migration admin.
+    error UnauthorizedMigration();
+
+    /// @notice Emitted when an account (whose status is being set) is 0x0.
+    error ZeroAccount();
+
+    /// @notice Emitted in constructor if Migration Admin is 0x0.
+    error ZeroMigrationAdmin();
+
+    /// @notice Emitted in constructor if Registrar is 0x0.
+    error ZeroRegistrar();
+
+    /* ============ Interactive Functions ============ */
+
+    /**
+     * @notice Sets the status for `account` to `status`.
+     * @notice If approving an earner that is already earning, but was recently removed from the Registrar earners list,
+     *         call `wrappedM.stopEarning(account)` before calling this, then call `wrappedM.startEarning(account)`.
```

Comment:
Is this required because `hasEarnerDetails` wouldn't be set otherwise?
Does that mean that the earner would still earn but no fee would be taken from the yield since `hasEarnerDetails` will be false?

Reply:
Correct. Earning WM is a stateful status that needs to be explicitly stopped. Directing a fee to the admin is a "claim time" process based on the details at that time.